enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,486 @@
+extended to add a shared CPUSet to the [Performance Profile CPU configuration](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpu)
+in addition to the existing reserved and isolated CPUSets. For example:
+
+```yaml
+cpu:
+  reserved: 0,1
+  shared: 2-5
+  isolated: 6-15
+```
+When the shared CPUSet is configured, NTO will:
+* Continue to set the systemd.cpu_affinity to the reserved CPUSet.
+* Apply kernel configuration for isolated CPUs (e.g. nohz_full) to only the isolated CPUSet.
+* Update kubelet configuration to specify that the isolated CPUSet is to be used for guaranteed QoS
+containers using whole CPUs.
+* Update OpenShift Kubernetes API Server configuration to enable a new admission hook. Note: although this would
+work in single-node deployments, we still need to work through how enablement will work in multi-node deployments
+(waiting to decide on design approach before working that out).
+
+The Kubelet configuration will be updated to allow a new CPUSet to be specified for guaranteed Qos containers
+using whole CPUs. There are two options:
+1. If this feature can be upstreamed, the [Kubelet Configuration](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) could be extended with a new `guaranteedCPUs` option.
+2. If this cannot be upstreamed, a new configuration file could be created and read by kubelet on startup - similar
+to the /etc/kubernetes/openshift-workload-pinning file created for The [Workload Partitioning](https://docs.openshift.com/container-platform/4.13/scalability_and_performance/ztp_far_edge/ztp-reference-cluster-configuration-for-vdu.html#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu) feature.
+
+The Kubelet (mostly the [CPU manager](https://github.com/openshift/kubernetes/tree/master/pkg/kubelet/cm/cpumanager) and its static policy) would be updated to:
+* Read and store the new `guaranteedCPUs` configuration.
+* Remove the `guaranteedCPUs` from the `defaultCpuSet` so that shared containers are excluded from the `guaranteedCPUs` when their affinity is set.
+* Extend the cpu_manager_state to include a new `guaranteedCpuSet` tracking the available CPUs.
+* When a guaranteed QoS container using whole CPUs is created, allocate CPUs from the `guaranteedCpuSet` instead 
+of the `defaultCpuSet` and set the CPU affinity to match.
+
+The next piece is to let the Kubernetes Scheduler take the split shared and guaranteed CPUSets into account when
+scheduling pods to a particular node. Kubernetes currently has a single 
+[CPU resource](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu) 
+that can be specified for each container. A change is necessary to ensure the scheduler will not over-allocate 
+shared or guaranteed CPUs on a node, resulting in a pod that would be scheduled but fail to run.
```

Comment:
Do we have any additional detail on how to intend to solve the scheduler visibility?

Reply:
nvm, I see it described below.